### PR TITLE
handle one biosample with multiple models efficiently

### PR DESCRIPTION
**Pull request by mayasheth**
edit "model application" workflow so that one biosample can have predictions made for multiple models if necessary
- generate features per biosample instead of per biosample/model pair
- output predictions in /model_name directory, which also allows you to confirm correct model was chosen

during model training, output files for shap score analysis

tested with a dryrun and biosample_config with two models for the same config

### CODE REVIEW

**Comment by mayasheth** [comment]
@atancoder Can you help with resolving the conflicts?

File: workflow/rules/feature_analysis.smk [hunk 1]
```diff
@@ -159,3 +159,17 @@ rule plot_permutation_feature_importance:
 	script:
 		"../scripts/model_training/plot_pfi.R"
 
+# rule plot_shape_scores:
```
Comment (atancoder) [inline workflow/rules/feature_analysis.smk:162]: Should this code be deleted?
s/shape/shap

File: workflow/rules/predictions.smk [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
+	input:
+		biosample_config = config["ABC_BIOSAMPLES_MODELS"],
+	output:
+		biosample_features = os.path.join(RESULTS_DIR, "{biosample}", "feature_table.tsv")
+	params:
+		e2g_path = config["E2G_DIR_PATH"]
+	conda:
+		"../envs/encode_re2g.yml"
+	resources:
+		mem_mb=8*1000
```
Comment (atancoder) [inline workflow/rules/predictions.smk:12]: probably only need 4GB here

File: workflow/scripts/model_training/plot_sffs.R [hunk 3]
```diff
@@ -16,10 +16,10 @@ feature_table = fread(feature_table_file)
 polynomial = gsub(" ", "", polynomial)
 
 # if polynomial==False, use nice_name for plotting
-if (polynomial=="FALSE" || polynomial == "False" || !polynomial){
+if (polynomial=="FALSE" || polynomial == "False"l){
```
Comment (atancoder) [inline workflow/scripts/model_training/plot_sffs.R:19]: typo?

**Review by atancoder — COMMENTED** [review]
Can you share what an example biosample config would look like with this new format? i.e can you show the configs you used when testing 1 ABC + 2 models?

**Comment by mayasheth** [comment]
> Can you share what an example biosample config would look like with this new format? i.e can you show the configs you used when testing 1 ABC + 2 models?

Biosample config is structured the same way.  See attached: in this case predictions are made for biosample K562_chr22 for both the model chosen by default (dhs_megamap, uploaded to the repo) and a duplicated version (dhs_megamap_dup) but only one set of ABC predictions are generated. 

[config_biosamples_test_dup.txt](https://github.com/EngreitzLab/ENCODE_rE2G/files/15314272/config_biosamples_test_dup.txt)

**Comment by mayasheth** [comment]
@atancoder check out update based on your suggestion? i still write an expanded biosamples config each time the Snakefile is run (needed as input to create a feature table per biosample) but make this a parameter so it shouldn't force a rerun.

File: workflow/rules/predictions.smk [hunk 4]
```diff
@@ -1,11 +1,18 @@
 
-rule write_feature_table: # write feature table to output directory for future reference
-    input:
-        feature_table_file = lambda wildcards: get_feature_table_file(wildcards.biosample)
-    output:
-        feature_table_here = os.path.join(RESULTS_DIR, "{biosample}", "feature_table.tsv")
-    run:
-        shell("cat {input.feature_table_file} > {output.feature_table_here}")
+rule make_biosample_feature_table:  # make feature table per biosample
```
Comment (atancoder) [inline workflow/rules/predictions.smk:2]: Do we still need this rule? Where is it being used? If we don't need it, we can remove creating the `ABC_BIOSAMPLES_MODELS` variable and file?

File: workflow/rules/predictions.smk [hunk 5]
```diff
@@ -27,15 +34,15 @@ rule generate_e2g_predictions:
 		final_features = os.path.join(RESULTS_DIR, "{biosample}", "genomewide_features.tsv.gz"),
 	params:
 		epsilon = config["epsilon"],
-		feature_table_file = os.path.join(RESULTS_DIR, "{biosample}", "feature_table.tsv"),
-		trained_model = lambda wildcards: get_trained_model(wildcards.biosample),
+		feature_table_file = lambda wildcards: os.path.join(str(BIOSAMPLE_DF.loc[(BIOSAMPLE_DF["biosample"]==wildcards.biosample) & (BIOSAMPLE_DF["model_dir_base"]==wildcards.model_name), "model_dir"]), "feature_table.tsv"),
+		trained_model = lambda wildcards: os.path.join(str(BIOSAMPLE_DF.loc[(BIOSAMPLE_DF["biosample"]==wildcards.biosample) & (BIOSAMPLE_DF["model_dir_base"]==wildcards.model_name), "model_dir"]), "model.pkl"),
```
Comment (atancoder) [inline workflow/rules/predictions.smk:38]: These look really complicated. Are you not able to utilize the existing `get_feature_table_file` function in utils.smk? We should be able to say something like
`feature_table_file = lambda wildcards: get_feature_table_file(wildcards.biosample, wildcards.model_name)`. And perhaps that function will implement what you have here

same goes with getting the model and thresholds

File: workflow/rules/utils.smk [hunk 6]
```diff
@@ -66,16 +69,13 @@ def _get_biosample_model_dir(biosample):
 	else:
 		return os.path.join(MODEL_DIR, f"{access_type}_intact_hic")
 
-def get_feature_table_file(biosample):
-	model_dir = _get_biosample_model_dir(biosample)
+def get_feature_table_file(model_dir):
```
Comment (atancoder) [inline workflow/rules/utils.smk:72]: Did these functions need to be changed? The behavior of these are so simple, they don't even need to be functions. See one of my comments above on how to utilize these functions

File: workflow/Snakefile [hunk 7]
```diff
@@ -47,6 +51,35 @@ BIOSAMPLE_ACTIVITES = {
...
+            new_row = row.copy()
+            new_row['model_dir'] = model_dir.strip()
+            new_rows.append(new_row)
+        return pd.DataFrame(new_rows)
+    else:
+        return pd.DataFrame([row])
+
+new_rows = BIOSAMPLE_DF.apply(split_rows, axis=1)
+BIOSAMPLE_DF = pd.concat(BIOSAMPLE_DF.apply(split_rows, axis=1).tolist(), ignore_index=True)
+BIOSAMPLE_DF = BIOSAMPLE_DF.drop_duplicates()
```
Comment (atancoder) [inline workflow/Snakefile:76]: when would we have duplicates?

File: workflow/Snakefile [hunk 8]
```diff
@@ -47,6 +51,35 @@ BIOSAMPLE_ACTIVITES = {
 # defined to be compatible with training workflow, which can start from an existing ABC directory
 ABC_BIOSAMPLES_DIR = {biosample: os.path.join(RESULTS_DIR, biosample) for biosample in BIOSAMPLE_DF.biosample}
 
+# add model directory path to biosamples df
```
Comment (atancoder) [inline workflow/Snakefile:54]: Some cleaner code suggestions:
1. We'd want this file to stay simple. It's better to move this logic into the utils.smk file. And in this file, we can just call something like `expand_biosamples_df(BIOSAMPLE_DF)`

2. We can combine all the logic (line 54-81) into 1 function that looks like the following:
```
def expand_biosample_df(biosample_df):
  new_rows = []
  for row in biosample_df:
    if model_dir not in row:
        row['model_dir'] = infer model_dir
    
    for model_dir in row['model_dir'].split(','):
      new_row = row.copy()
      new_row['model_dir'] = model_dir
      new_row['model_base'] = ..
      new_row['threshold'] = ..
      new_rows.append(new_row)
  return pd.DataFrame(new_rows)  
```

**Review by atancoder — COMMENTED** [review]
I think the overall structure looks good. Mostly some code improvement suggestions. Feel free to start testing this in scE2G while we hammer them out. 

If you can also update the README with the new output directory structure. If you rebase, I added some info about where the outputs are located, so you'd just need to change that to reflect the final prediction file path

File: workflow/rules/predictions.smk (hunk 4 above)
Comment (mayasheth) [reply workflow/rules/predictions.smk:2]: yes, this is needed so that the genome wide feature table contains features for all the models being applied

**Comment by mayasheth** [comment]
Made changes based on your code suggestions and edited the README of the main branch with some details on the new structure and links for the extended model feature tables!

File: workflow/rules/crispr_features.smk [hunk 9]
```diff
@@ -4,6 +4,7 @@ rule make_dataset_feature_table:
 	output:
 		dataset_features = os.path.join(RESULTS_DIR, "{dataset}", "feature_table.tsv")
 	params:
+        model_config = config["model_config"],
```
Comment (atancoder) [inline workflow/rules/crispr_features.smk:7]: the formatting here looks weird

File: workflow/rules/predictions.smk [hunk 10]
```diff
@@ -1,11 +1,18 @@
...
-        feature_table_here = os.path.join(RESULTS_DIR, "{biosample}", "feature_table.tsv")
-    run:
-        shell("cat {input.feature_table_file} > {output.feature_table_here}")
+rule make_biosample_feature_table:  # make feature table per biosample
+	input:
+		config["ABC_BIOSAMPLES"]
+	output:
+		biosample_features = os.path.join(RESULTS_DIR, "{biosample}", "feature_table.tsv")
+	params:
+		biosample_config = config["ABC_BIOSAMPLES_MODELS"],
```
Comment (atancoder) [inline workflow/rules/predictions.smk:8]: It seems like you really just need the model directories. You can just leverage the existing BIOSAMPLES_DF to get that. Something like 
```
model_dirs = BIOSAMPLE_DF[BIOSAMPLE_DF['biosample'] == biosample]['model_dir'].to_list()
```
I believe snakemake and R should handle model_dirs being a list, but not completely sure. 

Then this way, you don't need to create a separate config file in Snakemake. (It also looks like you may've forgotten to create the ABC_BIOSAMPLES_MODELS config in the latest commit?)

**Review by atancoder — APPROVED** [review]
This looks pretty good. Thanks for cleaning it up! 
Before merging into main, please run the pipeline through a test run just to make sure there's no errors

**Pull request merged**